prover/src/constraints/evaluator.rs

```diff
@@ -157,6 +158,21 @@ impl<'a, A: Air, E: FieldElement<BaseField = A::BaseField>> ConstraintEvaluator<
         // LDE domain
         let lde_shift = domain.ce_to_lde_blowup().trailing_zeros();
 
+        // we create two maps: one for storing the adjustments x^i for all needed i
```

Comment:
Trees are used but they are not actually needed. Each group has its own (unique) degree adjustment. Nevertheless, this is a minuscule overhead and it can be useful in more complex cases. It is actually necessary in custom divisors where one considers a different constraint grouping strategy.

Reply:
I've been wondering about this but given the improvement was thinking I was wrong! If we're sure we will adopt this strategy for custom divisors, we could keep it as is, otherwise let's remove it so that only what's strictly beneficial for now appears.

Reply:
This grouping (per degree) does only apply because there is only one divisor. I also think that maps will be needed with boundary constraints, I will verify this. If it is not needed there we can modify this for now. But in any case I suspect the effect will be small in all cases so I would go for the "cleanest" solution.

Reply:
Regarding use of maps: currently if the same degree is found in a main and an aux or boundary constraint it is not recomputed. Removing maps will need either re-computation or more complex book-keeping (e.g. processing groups in a sorted order). However, since constraints are evaluated by first main, then aux and finally boundary this can get complex. It seems this is too much trouble for little gain. Maps can easily be removed from the verifier side, I will include this in a follow-up commit (this seems to not have a noticeable effect though).